community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/idmapping/string/EncodingIdMapper.java

```diff
@@ -452,9 +450,50 @@ private void buildCollisionInfo( InputIterator<Object> ids, ProgressListener pro
     private void detectDuplicateInputIds( Radix radix, NumberArrayStats collisionDataCacheStats,
             List<String> sourceDescriptions ) throws InterruptedException
     {
+        // We do this collision sort using ParallelSort which has the data cache and the tracker cache,
+        // the tracker cache gets sorted, data cache stays intact. In the collision data case we actually
+        // have one more layer in here so we have tracker cache pointing to collisionDataIndexCache
+        // pointing to dataCache. This can be done using the ParallelSort.Comparator abstraction.
+        //
+        // The Comparator below takes into account dataIndex for each eId its comparing so that an extra
+        // comparison based on dataIndex is done if it's comparing two equal eIds. We do this so that
+        // stretches of multiple equal eIds are sorted by dataIndex order, to be able to write an efficient
+        // duplication scanning below and to have deterministic duplication reporting.
+        Comparator comparator = new Comparator()
+        {
+            @Override
+            public boolean lt( long left, long pivot )
+            {
+                long leftEId = dataCache.get( left );
+                long pivotEId = dataCache.get( pivot );
+                if ( DEFAULT.lt( leftEId, pivotEId ) )
+                {
+                    return true;
+                }
+                if ( leftEId == pivotEId )
+                {
+                    return left < pivot;
+                }
+                return false;
+            }
+
+            @Override
+            public boolean ge( long right, long pivot )
+            {
+                long rightEId = dataCache.get( right );
+                long pivotEId = dataCache.get( pivot );
+                if ( DEFAULT.ge( rightEId, pivotEId ) )
+                {
+                    return rightEId == pivotEId ? right > pivot : true;
```

Comment:
correct, although there's no chance that they're ever the same actually. They just can't be. It should be an assertion if anything.